src/Okta.Sdk/DefaultRetryStrategy.cs

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2014 - present Okta, Inc. All rights reserved.
+// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Okta.Sdk
+{
+    /// <summary>
+    /// Default retry strategy
+    /// </summary>
+    public class DefaultRetryStrategy : IRetryStrategy
+    {
+        /// <summary>
+        /// The default delta used in the back-off formula to account for some clock skew in our service
+        /// </summary>
+        public static readonly int DefaultBackOffSecondsDelta = 1;
+
+        /// <inheritdoc/>
+        public int MaxRetries { get; set; }
+
+        /// <inheritdoc/>
+        public int RequestTimeOut { get; set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DefaultRetryStrategy"/> class.
+        /// </summary>
+        /// <param name="maxRetries">the number of times to retry</param>
+        /// <param name="requestTimeOut">The request timeout in seconds</param>
+        public DefaultRetryStrategy(int maxRetries, int requestTimeOut)
+        {
+            MaxRetries = maxRetries;
+            RequestTimeOut = requestTimeOut;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DefaultRetryStrategy"/> class.
+        /// </summary>
+        public DefaultRetryStrategy()
+        {
+        }
+
+        /// <inheritdoc/>
+        public async Task<HttpResponseMessage> WaitAndRetryAsync(HttpRequestMessage request, Func<HttpRequestMessage, Task<HttpResponseMessage>> operation)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            if (operation == null)
+            {
+                throw new ArgumentNullException(nameof(operation));
+            }
+
+            var attempts = 0;
+            var stopwatch = new Stopwatch();
+            stopwatch.Start();
+
+            do
+            {
+                attempts++;
+
+                var response = await operation(request).ConfigureAwait(false);
+                var statusCode = (int)response.StatusCode;
+
+                try
+                {
+                    if (statusCode == 429 && attempts < MaxRetries + 1 &&
+                        ((RequestTimeOut > 0 && stopwatch.Elapsed.Seconds < RequestTimeOut) || RequestTimeOut == 0))
+                    {
+                        await CreateDelayAsync(response).ConfigureAwait(false);
+                        response.Headers.TryGetValues("X-Okta-Request-Id", out var requestId);
+                        AddRetryOktaHeaders(request, requestId.First(), attempts);
+                    }
+                    else
+                    {
+                        return response;
+                    }
+                }
+                catch
```

Comment:
I don't think I understand the purpose of this `try...catch`. The code inside this block all looks like internal logic and I don't see any situations where it is expected to throw. So if it *did* throw it would be a bug and we definitely would not want to swallow the exception. Also there are exceptions like `OperationCanceledException` or `OutOfMemoryException` that should never be swallowed.

It seems like we should just remove the `try...catch` unless I'm missing something.

Reply:
I throw an `InvalidOperationException`  when we have multiple `X-Rate-Limit-Reset` headers. 
For those case, we shouldn't retry and just return the error.

Reply:
If what you want is to return the original `response` in this case (and also if the backoff time exceeds the request timeout), I think it would be good to write the code in a way that doesn't require the SDK to throw and immediately catch an exception.

Throwing an exception takes a lot of system resources, so its best practice not to throw an exception and immediately catch it just to implement control flow. It's also confusing because people don't expect a class to throw an exception in one method then catch it in another, and will automatically assume the exceptions are being thrown to the caller (I made this assumption until you explained the logic).

Also there are cases where .NET library methods throw `InvalidOperationException` and you would not want to accidentally catch those.

Maybe rewrite `CreateDelayAsync` so that it returns a `TimeSpan` instead of a Task, and if it returns `TimeSpan.Zero` then you break out of the loop.

Reply:
I'm also wondering why we stop retrying if we see two or more `X-Rate-Limit-Reset` headers. It seems like there are more graceful options for handling that situation:
1. Ignore the `X-Rate-Limit-Reset` header and fallback to using the logic that applies when that header isn't present.
2. Choose one of the values to use (i.e. always use the first one).